### PR TITLE
Module7: Server Side Rendering

### PR DESCRIPTION
**Pull request by nkorczyk**
Changes proposed in this PR:

Implemented Server Side Rendering.
Added necessary libraries.
Webpack configuration for client & server.
Route Masking.

Please check and review.

### CODE REVIEW

File: src/serverRenderer.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import SearchPage from './components/SearchPage';
+import MoviePage from './components/MoviePage';
+import NotFoundPage from "./components/NotFoundPage";
+import configureStore from './actions/store';
+
+const routes = [
+  { path: '/', exact: true, component: SearchPage },
+  { path: '/search/:query', exact: true, component: SearchPage },
+  { path: '/film/:id', exact: true, component: MoviePage },
+  { path: '/*', component: NotFoundPage }
```
Comment (MichalWierzba) [inline src/serverRenderer.js:16]: Routes configuration is now here and in `Root` component. Try removing this duplication.

File: src/serverRenderer.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+export default function serverRenderer() {
+  return (req, res) => {
+    const store = configureStore();
+    const branch = matchRoutes(routes, req.url);
+    const promises = branch.map(({ route, match }) => {
+      const { fetchData } = route.component;
+
+      if (!(fetchData instanceof Function)) {
```
Comment (MichalWierzba) [inline src/serverRenderer.js:47]: `fetchData` does not exist anywhere, so data is not fetched on server. Currently empty markup is served from server. Please fix.

**Review by MichalWierzba — CHANGES_REQUESTED** [review]
Code is well organized and looks really good. However there is one issue with data fetching causing SSR not working as expected. Please check comment for details.

**Comment by MichalWierzba** [comment]
There is strange issue on Movie page - infinite loop of server requests. Please check:
image

**Comment by nkorczyk** [comment]
Infinite loop bug fixed.

**Comment by nkorczyk** [comment]
I've made changes related route masking. When I type `url` like: http://localhost:8000/film/460790
in `view source` I see:




And for url like: http://localhost:8000/search/star
in `view source` I see:
star

File: src/serverRenderer.js (hunk 2 above)
Comment (nkorczyk) [reply src/serverRenderer.js:47]: Data is now fetched from the server. Markup is fulfilled.

File: package.json [hunk 3]
```diff
@@ -5,10 +5,9 @@
   "entry point": "srcServer.js",
   "main": "script.js",
   "scripts": {
-    "start": "babel-node buildScripts/srcServer.js",
-    "open:src": "webpack-dev-server --hot --port 80",
-    "build:dev": "webpack --env dev --progress --profile --colors",
-    "build:prod": "webpack --env prod --progress --profile --colors",
+    "start": "cross-env NODE_ENV=development  node ./buildScripts",
+    "build": "cross-env NODE_ENV=production webpack --config ./webpack",
+    "prod": "cross-env NODE_ENV=production  node ./buildScripts",
```
Comment (MichalWierzba) [inline package.json:10]: Please check, if `prod` script works properly.

File: src/components/Content.jsx [hunk 4]
```diff
@@ -1,18 +1,32 @@
...
-  )
+class Content extends Component {
+
+  static fetchData(dispatch, match) {
+    return dispatch(searchMovies(match.params.query));
+  }
+
+  componentDidMount() {
+    if (this.props.match) {
+      this.props.searchMovies(this.props.match.params.query);
```
Comment (MichalWierzba) [inline src/components/Content.jsx:17]: Please make sure `serachMovies` is not run on initial render on client side, when it was already done on server side.

**Review by MichalWierzba — APPROVED** [review]

**Pull request merged**